components/utilities/floatingNav.js

```diff
@@ -1,204 +1,145 @@
-      const top = hrefs[0].offsetTop;
-      if (top < highest || highest === 0) {
-        highest = top;
-        if (hrefs.length > 0) {
-          const link = hrefs[0].getAttribute("href");
-          this.setState({ target: link });
-        }
-      }
-    }
-  }
-
-  async generateMenu() {
-    if (this.state.headers.length > 0) {
-      this.closeMenu();
-    }
-    const tocMenu = [];
-    const headers = Array.prototype.slice.apply(
+  useEffect(() => {
+    const headingElements = Array.from(
       document.querySelectorAll(
         "article.leaf-page h1, article.leaf-page h2, article.leaf-page h3, article.leaf-page h4, article.leaf-page h5, article.leaf-page h6"
       )
     );
-    const observer = this.state.observer;
-    for (const index in headers) {
-      const ele = headers[index];
-      if (ele.getElementsByTagName === undefined) {
-        continue;
-      }
-      const hrefs = ele.getElementsByTagName("a");
-      if (hrefs.length > 0) {
-        const target = hrefs[0].getAttribute("href");
-        tocMenu.push({
-          label: ele.innerText,
-          target: target,
-          level: ele.tagName,
-        });
-        observer.observe(ele);
-      }
-    }
-    this.setState({ menu: tocMenu, headers: headers });
-  }
 
-  closeMenu() {
-    for (const index in this.state.headers) {
-      const ele = this.state.headers[index];
-      this.state.observer.unobserve(ele);
-    }
-    this.setState({ menu: [], headers: [] });
-  }
+    const newNestedHeadings = getNestedHeadings(headingElements);
+    setNestedHeadings(newNestedHeadings);
+  }, []);
 
-  async componentDidMount() {
-    const observer = new IntersectionObserver(
-      this.handleIntersection,
-      this.state.options
-    );
-    await this.setState({ observer: observer, menu: [], headers: [] });
-    this.props.router.events.on("routeChangeComplete", this.handleRouteChange);
-    window.addEventListener("ChangeTheme", this.handleTheme);
+  return { nestedHeadings };
+};
 
-    this.generateMenu();
-  }
+const getNestedHeadings = (headingElements) => {
+  const nestedHeadings = [];
 
-  handleRouteChange() {
-    this.closeMenu();
-    this.generateMenu();
+  for (const index in headingElements) {
+    const ele = headingElements[index];
+    if (ele.getElementsByTagName === undefined) {
+      continue;
+    }
+    const hrefs = ele.getElementsByTagName("a");
+    if (hrefs.length > 0) {
+      const target = hrefs[0].getAttribute("href");
+      nestedHeadings.push({
+        label: ele.innerText,
+        target: target,
+        level: ele.tagName,
+      });
+    }
   }
 
-  componentWillUnmount() {
-    this.closeMenu();
-    window.removeEventListener("ChangeTheme", this.handleTheme);
-    this.props.router.events.off("routeChangeComplete", this.handleRouteChange);
-  }
+  return nestedHeadings;
+};
 
-  handleTheme() {
-    this.setState({ theme: document.body.dataset.theme });
-  }
+const useIntersectionObserver = (setActiveId) => {
```

Comment:
I don't really understand what this function is supposed to do. Can you add a docstring explaining?

Reply:
Makes total sense. Since this component was refactored almost entirely, I used a different logic to get the headings on the page and map them, but didn't properly explained it. Added the explanation on aa1e2c2